streams/src/main/java/org/apache/kafka/streams/state/internals/CachingSessionStore.java

```diff
@@ -164,30 +164,62 @@ public void put(final Windowed<Bytes> key, final byte[] value) {
         return findSessions(from, to, 0, Long.MAX_VALUE);
     }
 
-    private void putAndMaybeForward(final ThreadCache.DirtyEntry entry, final InternalProcessorContext context) {
+    private void putAndMaybeForward(final ThreadCache.DirtyEntry entry,
+                                    final InternalProcessorContext context) {
         final Bytes binaryKey = cacheFunction.key(entry.key());
         final ProcessorRecordContext current = context.recordContext();
         context.setRecordContext(entry.entry().context());
         try {
             final Windowed<K> key = SessionKeySchema.from(binaryKey.get(), serdes.keyDeserializer(), topic);
             final Bytes rawKey = Bytes.wrap(serdes.rawKey(key.key()));
+            final byte[] newValue = entry.newValue();
             if (flushListener != null) {
-                final AGG newValue = serdes.valueFrom(entry.newValue());
-                final AGG oldValue = newValue == null || sendOldValues ? fetchPrevious(rawKey, key.window()) : null;
-                if (!(newValue == null && oldValue == null)) {
+                final byte[] oldRawValue;
+                if (sendOldValues || newValue == null) {
+                    try (final KeyValueIterator<Windowed<Bytes>, byte[]> iterator =
+                             bytesStore.findSessions(rawKey, key.window().start(), key.window().end())) {
+                        if (!iterator.hasNext()) {
+                            oldRawValue = null;
+                        } else {
+                            oldRawValue = iterator.next().value;
+                        }
+                    }
+                } else {
+                    oldRawValue = null;
+                }
+                final FlushEntry<K, AGG> flushEntry = flushEntry(
+                    serdes,
+                    entry.key().get(),
+                    newValue,
+                    oldRawValue,
+                    entry.entry().context().timestamp());
```

Comment:
I'm sorry; getting a little cross-eyed here...

It looks like we're putting `newValue` and `oldRawValue` in a new `FlushEntry` instance here for the sole purpose of extracting them back out and discarding the `FlushEntry` object immediately following.

What am I missing here?

Reply:
> discarding the FlushEntry object immediately following.

We don't discard it. It's use the in next lines:

```
if (flushEntry.value != null || flushEntry.oldValue != null) {
flushListener.apply(
                        key,
                        flushEntry.value,
                        flushEntry.oldValue,
                        entry.entry().context().timestamp());
}
```

Maybe I miss-understand your question?

Reply:
Yeah, that's what I meant. We put the new and old value in, and then immediately take them back out again, and never do anything else with the `flushEntry`. It just seems like we could skip the middle-man and directly deserialize the values and pass them to the flushListener.

But it's also no big deal... I was just curious if I missed something.

Reply:
Well. As you mentioned, we pass in objects into `flushEntry()` and it creates a `FlushEntry` with serialized data. To share code, and to apply the same patter for all stores, it makes sense IMHO to push the serialization work into `flushEntry()` method -- the caller should be agnostic if `flushEntry` modified the data or not.